### PR TITLE
NH-3905 - Implement Async operations

### PR DESCRIPTION
**Pull request by fredericDelaporte**
[NH-3905](https://nhibernate.jira.com/browse/NH-3905) This is a rebase of #588.

For eliminating the different directions async future have taken before (and for easing the rebase), most future related changes have been squashed into first @maca88 commit.
For allowing to generates async files and compile most commits individually, most async generation setup commits have been moved prior to first maca88 commit. This means most upgrade async generator commits are squashed and the generator version is 0.3.6 from almost "the beginning" of this rebase.
Some additional squashing have been made (especially changes reverted some commits later).
Most "regen async" commits have been regenerated with 0.3.6 async generator version.

### CODE REVIEW

**Comment by hazzik** [comment]
@fredericDelaporte I think we need to move "Regenerate Async" to the latest commit. Made in one step.

**Comment by hazzik** [comment]
I did further combination of commits and moved async generation to the top. 

This is the diff to #588. In short less files need to be `partial`

```diff
diff --git a/build-common/common-project.xml b/build-common/common-project.xml
index d0dba772f..a075fd4c6 100644
--- a/build-common/common-project.xml
+++ b/build-common/common-project.xml
@@ -54,8 +54,8 @@
 			<arg value="install" />
 		</exec>
 	</target>
-
-	<target name="common.solution-restore" depends="common.nuget-restore">
+  
+  <target name="common.solution-restore" depends="common.nuget-restore">
 		<exec program="${path::combine(tools.dir, 'msbuild.cmd')}" verbose="true">
 			<arg value="/t:Restore" />
 			<arg value="${root.dir}/src/NHibernate.sln" />
diff --git a/src/NHibernate.Test/Async/NHSpecificTest/NH1332/Fixture.cs b/src/NHibernate.Test/Async/NHSpecificTest/NH1332/Fixture.cs
index bdfd1d361..d5f7ed1a7 100644
--- a/src/NHibernate.Test/Async/NHSpecificTest/NH1332/Fixture.cs
+++ b/src/NHibernate.Test/Async/NHSpecificTest/NH1332/Fixture.cs
@@ -18,7 +18,7 @@ namespace NHibernate.Test.NHSpecificTest.NH1332
 	using System.Threading.Tasks;
 	using System.Threading;
 	[TestFixture]
-	public class FixtureAsync : BugTestCase
+	public partial class FixtureAsync : BugTestCase
 	{
 		private static readonly ILog log = LogManager.GetLogger(typeof(FixtureAsync));
 
@@ -60,7 +60,7 @@ protected override void Configure(Configuration configuration)
 			}
 		}
 
-		public class PostCommitDelete : IPostDeleteEventListener
+		public partial class PostCommitDelete : IPostDeleteEventListener
 		{
 			public Task OnPostDeleteAsync(PostDeleteEvent @event, CancellationToken cancellationToken)
 			{
diff --git a/src/NHibernate.Test/Async/NHSpecificTest/NH1882/TestCollectionInitializingDuringFlush.cs b/src/NHibernate.Test/Async/NHSpecificTest/NH1882/TestCollectionInitializingDuringFlush.cs
index f6d572abd..620d60935 100644
--- a/src/NHibernate.Test/Async/NHSpecificTest/NH1882/TestCollectionInitializingDuringFlush.cs
+++ b/src/NHibernate.Test/Async/NHSpecificTest/NH1882/TestCollectionInitializingDuringFlush.cs
@@ -20,7 +20,7 @@ namespace NHibernate.Test.NHSpecificTest.NH1882
 	using System.Threading.Tasks;
 	using System.Threading;
 	[TestFixture]
-	public class TestCollectionInitializingDuringFlushAsync : TestCaseMappingByCode
+	public partial class TestCollectionInitializingDuringFlushAsync : TestCaseMappingByCode
 	{
 		private readonly InitializingPreUpdateEventListener listener = new InitializingPreUpdateEventListener();
 
@@ -63,7 +63,7 @@ protected override HbmMapping GetMappings()
 			return mapper.CompileMappingForAllExplicitlyAddedEntities();
 		}
 
-		public class InitializingPreUpdateEventListener : IPreUpdateEventListener
+		public partial class InitializingPreUpdateEventListener : IPreUpdateEventListener
 		{
 			public static InitializingPreUpdateEventListener Instance = new InitializingPreUpdateEventListener();
 
diff --git a/src/NHibernate.Test/NHSpecificTest/NH1710/Fixture.cs b/src/NHibernate.Test/NHSpecificTest/NH1710/Fixture.cs
index 260ed30d3..100032f5c 100644
--- a/src/NHibernate.Test/NHSpecificTest/NH1710/Fixture.cs
+++ b/src/NHibernate.Test/NHSpecificTest/NH1710/Fixture.cs
@@ -11,7 +11,7 @@ public class A
 		public virtual decimal? Amount { get; set; }
 	}
 
-	public abstract class BaseFixture
+	public abstract partial class BaseFixture
 	{
 		protected const string TestNameSpace = "NHibernate.Test.NHSpecificTest.NH1710.";
 		protected Configuration cfg;
diff --git a/src/NHibernate/Impl/EnumerableImpl.cs b/src/NHibernate/Impl/EnumerableImpl.cs
index f26870d74..d554a1784 100644
--- a/src/NHibernate/Impl/EnumerableImpl.cs
+++ b/src/NHibernate/Impl/EnumerableImpl.cs
@@ -21,7 +21,7 @@ namespace NHibernate.Impl
 	/// will cause it to be disposed, probably unexpectedly for the developer. (https://stackoverflow.com/a/11179175/1178314)
 	/// "Fortunately", it does not currently support multiple iterations anyway.</para>
 	/// </remarks>
-	public class EnumerableImpl : IEnumerable, IEnumerator, IDisposable
+	public partial class EnumerableImpl : IEnumerable, IEnumerator, IDisposable
 	{
 		private static readonly IInternalLogger log = LoggerProvider.LoggerFor(typeof(EnumerableImpl));
 
diff --git a/src/NHibernate/NHibernate.csproj.DotSettings b/src/NHibernate/NHibernate.csproj.DotSettings
new file mode 100644
index 000000000..7b3bb7c08
--- /dev/null
+++ b/src/NHibernate/NHibernate.csproj.DotSettings
@@ -0,0 +1,2 @@
+﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:Boolean x:Key="/Default/CodeInspection/NamespaceProvider/NamespaceFoldersToSkip/=async/@EntryIndexedValue">True</s:Boolean></wpf:ResourceDictionary>
\ No newline at end of file

```

**Comment by fredericDelaporte** [comment]
> I think we need to move "Regenerate Async" to the latest commit. Made in one step.

Thought the same at first but this means previous commits cannot compile without first generating async files, modifying the tree. It may not be needed to be able to compile previous commits but I still rather avoid letting many non compilable commits in history. Maybe just one regen right after remaining maca88 commit would have been enough for this purpose though.

**Comment by hazzik** [comment]
Then we, probably, need to clash all these prep commits into one.

**Comment by fredericDelaporte** [comment]
Move most prep prior to async generation commit, done some additional squashing, dropped syntax workaround for async gen no more needed, cleaned some undue added namespace, split maca88 commit between prep and what is requiring to already have async files generating (Linq & Future).

This yield an intial async generation, followed by changes for exploiting it for linq & futures, followed by a regen.

**Comment by hazzik** [comment]
All good. I think we might need to move `Future` refactoring into the future ticket (we had one for 5.0)

File: src/NHibernate/Criterion/QueryOver.cs [hunk 1]
```diff
@@ -102,10 +102,10 @@ private IFutureValue<U> FutureValue<U>()
...
 		}
 
-		/// <summary>
-		/// Get an executable instance of <c>IQueryOver&lt;TRoot&gt;</c>,
-		/// to actually run the query.</summary>
-		public IQueryOver<TRoot,TRoot> GetExecutableQueryOver(ISession session)
+        /// <summary>
+        /// Get an executable instance of <c>IQueryOver&lt;TRoot&gt;</c>,
+        /// to actually run the query.</summary>
+        public IQueryOver<TRoot,TRoot> GetExecutableQueryOver(ISession session)
```
Comment (fredericDelaporte) [inline src/NHibernate/Criterion/QueryOver.cs]: Whitespace miss probably introduced during rebasing. I am anyway rebasing again, while rewording commits for including Jira number.

**Comment by hazzik** [comment]
I'm going to merge as soon as tests pass. Any objections?

**Comment by fredericDelaporte** [comment]
Ok for me, going to do that myself in a couple of minutes if you do not seem online.

**Pull request merged**